### ISSUE TITLE
If an express router handles a request, `onSend` hooks are not called

### ISSUE THREAD

**Issue by SimenB** [fastify/fastify-express#20]
## 🐛 Bug Report

We're trying to migrate from `express-session` to `fastify-session`. It mostly seems to be working except for the cookie-setting (which is not happening), meaning our session only lives as long as the request.

I've narrowed the issue down (I think) to `fastify-session` using `onSend` to set the cookie: https://github.com/SerayaEryn/fastify-session/blob/e201f78fc9d7bd33c6f2e84988be7c8af4b5a8a3/lib/fastifySession.js#L25

And I believe the issue is that we have a `express.Router` handling the request from the client that ends up setting a session, and then doing a redirect. Since an `express` handler handles the request, I believe `onSend` is not called, and no `Set-Cookie` header is provided in the response.

## To Reproduce

```js
const Fastify = require('fastify');
const { Router } = require('express');

async function build() {
  const fastify = Fastify();
  fastify.addHook('onSend', (req, res, body, next) => {
    console.log('on send called!');
    next();
  });

  await fastify.register(require('fastify-express'));

  const router = Router();

  router.get('/expr', (_, res) => {
    res.json({ hello: 'world' });
  });

  fastify.use(router);

  fastify.get('/fast', () => {
    return { hello: 'fast' };
  });

  return fastify;
}

build()
  .then((fastify) => fastify.listen(4000))
  .catch(console.log);
```

If you do `curl http://localhost:4000/fast` you'll see a `on send called!` logged out, but doing `curl http://localhost:4000/expr` will not print that

## Expected behavior

I'd expect `onSend` to still be called even though an express router has handled the request.

While this code (mixing express and fastify) is temporary, I'd really like to be able to land migration from `express-session` to `fastify-session` without also having to migrate all our handlers at the same time.

In theory we can probably try to port `express-session`'s [hacky `res.end` replacement](https://github.com/expressjs/session/blob/a8641429502fcc076c4b2dcbd6b2320891c1650c/index.js#L250), but that feels super brittle... Although I guess the contract of "all `express` handlers and middleware run in `onRequest` phase - if request is handled no further hooks are called". If that's the case, some help coming up with a workaround would be lovely. (can be hacky as longs as it's not too brittle - just need to be able to land pieces of migration rather than all at once)

## Your Environment

- *node version*: 14.15.3
- *fastify version*: 3.12.0
- *os*: Mac

**Comment by mcollina** [comment]
TL;DR - there is no elegant solution to monkeypatching.

Unfortunately much of the express ecosystem relies on monkey patching the `end()` method. This is one of the key technical differences between the two frameworks, and partially one of the reasons why @delvedor and myself started Fastify long ago.

In essence this issue is out of scope. As you said, this behavior is explicitly called in the README because we did not want to get into the monkey patching Express (that monkey patches node core), as it's going to be _messy_ and it won't work for all applications so it would need to stay behind an option, disabled by default. The good thing is that I think it's doable.

**Comment by SimenB** [comment]
>The good thing is that I think it's doable.

Happy to take a stab at a PR if you can provide some guidance on where/how to approach this 🙂

**Comment by mcollina** [comment]
The quickest way is probably to lift https://github.com/SerayaEryn/fastify-session/blob/e201f78fc9d7bd33c6f2e84988be7c8af4b5a8a3/lib/fastifySession.js#L93-L113 and do the monkey patching `express-session` does before fastify-express is set up. You might also have to fork fastify-express to do this - as I said it's not going to be nice.

We might also add an option function that will be called to do these kind of monkey-patching manually.

**Comment by SimenB** [comment]
Thanks! So just to verify I've understood correctly

1. Before registering `express-fastify`, add the code from `express-session` which monkeypatches `res.end` so that I can override it and add a hook to do something:tm: before response is sent to the client
2. This _might_ require some changes in `fastify-express` as well, that is as of now unclear
3. Make sure to call `onSend` from `fastify-session` within the hook added in step 1

Does that seem right? And 1&2 might be accepted as PRs to this repo behind an option

**Comment by mcollina** [comment]
Yes exactly regarding the PRs. I would not make the new code/options specific to express-session, but make it pluggable (by calling a user-provided function).

**Comment by SimenB** [comment]
I've been swamped for the last few months. I do plan to get back to this some time in the summer

**Comment by flipperlite** [comment]
The following information may dissuade you from completing this task. There are other issues at play when mixing native Fastify and non-native Express routes in the same app. My recommendation is to create a brand new repo/app for your migration and not mix native/non-native routes.

I was trying to migrate an existing Express app putting the old and new routes into the same app. I'm sorry to report that the "monkeypatching" may be at play here too. There are two bad scenarios:

1. When an Express 404 handler is present, all valid, native Fastify routes are incorrectly caught by it resulting in 404.
2. When the Express 404 handler is removed all non-existent routes are correctly caught by the Fastify OOTB 404 handler and GET routes seem to work as shown in the OP. Unfortunately, with the PATCH route in the example below, the request just hangs because of the Express body parsers. I confirmed by commenting out the Express body parsers and the [PATCH no longer hangs but I no longer have access to req.body](https://github.com/fastify/fastify-express/issues/73#issuecomment-1067022626).

```js
const fastify = require('fastify')()
const express = require('express')
const router = express.Router()

// non-Express, native route. Caught by 404 hander if it exists, or hangs if 404 handler doesn't exist.
fastify.get('/hello-fastify', handleGet)
fastify.patch('/bar-fastify', handlePatch)

// Express routes
router.use(function (req, res, next) {
  res.setHeader('x-custom', true)
  next()
})

router.get('/hello', handleGet)

router.get('/foo', (req, res) => {
  res.status(400)
  res.json({ foo: 'bar' })
})

router.patch('/bar', handlePatch)

// Express 404 handler. When present, fastify routes get 404. When commented out, GET works but PATCH hangs.
// router.use('*', (req, res) => {
//   res.status(404)
//   res.json({ msg: 'not found'})
// })

fastify.register(require('fastify-express'))
  .after(() => {
    fastify.use(express.urlencoded({extended: false})) // for Postman x-www-form-urlencoded
    fastify.use(express.json())

    fastify.use(router)
  })

fastify.listen(3000, console.log)

async function handlePatch (req, res) {
  if (!req.body || Object.keys(req.body).length === 0) {
    res.status(400)
    res.json({ msg: 'no req.body'})
  } else {
    res.status(200)
    res.json(req.body)
  }
}

async function handleGet (req, res) {
  res.status(201)
  const output = { hello: 'world' }
  res.json ? res.json(output) : res.send(output)
}
```
My curl commands:
```bash
me@computer ~ % curl -X GET http://localhost:3000/asdf (CORRECT with Fastify OOTB 404 handler)
{"message":"Route GET:/asdf not found","error":"Not Found","statusCode":404}

me@computer ~ % curl -X GET http://localhost:3000/hello (CORRECT)
{"hello":"world"}

me@computer ~ % curl -X GET http://localhost:3000/hello-fastify (INCORRECT with Express 404 handler)
{"msg":"not found"}

me@computer ~ % curl -X GET http://localhost:3000/hello-fastify (CORRECT w/o Express 404 handler)
{"hello":"world"}

me@computer ~ % curl -X GET http://localhost:3000/asdf (CORRECT with Express 404 handler)
{"msg":"not found"}

me@computer ~ % curl -X PATCH -H 'content-type:application/json' -d '{"foo2":"bar2"}' http://localhost:3000/bar-fastify (HANGS w/o Express 404 handler)
^C

me@computer ~ % curl -X PATCH -H 'content-type:application/json' -d '{"foo2":"bar2"}' http://localhost:3000/bar-fastify (INCORRECT with Express 404 handler)
{"msg":"not found"}
```
